# Solution Combinatorics with Voiceovers
*Arthur Ryman, last updated 2025-04-14*

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/agryman/instant-insanity/blob/main/notebooks/Solution-Combinatorics-Voiceovers.ipynb)

## Introduction

This notebook focuses on the combinatorics of the Instant Insanity puzzle, namely how to count the number of combinations of the cubes.
Other notebooks will focus on other aspects of the puzzle such as its graph theory solution and history.

This notebook contains both voiceovers and explanatory text.
The voiceovers are intended for use in the video.
The explanatory text goes into more detail but may be too detailed for use in video.

## Change History

This is a working document and will be edited frequency.
I will annotate each voiceover with the date of its most recent update.

## Combinatorics

This section should come after the general introduction to the puzzle.
We should have already described the puzzle and defined what it means to solve it, namely you are required to line up the four cubes in a horizontal row such that
no colour is repeated along the front, back, top, or bottom sides.

This notebook shows how to compute the total number of combinations of the Instant Insanity puzzle.
The branch of mathematics that focuses on how to count combinations is called *Combinatorics*.
The number of combinations for any puzzle is very interesting since it indicates how hard it is to solve the
puzzle.

## How Difficult is This Puzzle?

### Scene

Show an animation that hints at the difficulty of searching for a solution.
Show the row of four cubes. Pick one cube at random and give it a random rotation.

### Voiceover: last updated 2025-04-11

How difficult is it to solve the puzzle?
One way to solve it is to try every possible combination of the cubes until we find one that works.
In computer science, this solution strategy is called a brute-force search.
In the absence of any further insight into the puzzle,
its difficulty is therefore directly proportional to the total number of combinations.

So how many combinations are there?

## 82,944 or 41,472?

### Scene

Show the text "82,944 or 41,472?"

### Voiceover: last updated 2025-04-11

Here we encounter an interesting situation. 
We have two competing claims for the number of combinations.
The puzzle package, and several YouTube videos, claim that there are 82,944 combinations.
However, an article published at Cambridge University in 1947 gives an elegant solution to the puzzle and claims there are only 41,472.

## Winning Solutions: 82,944

See: [Instant Insanity](https://winning-moves.com/product/InstantInsanity.asp)

### Scene

Show the puzzle package and the text: "82,944 Combinations. Can You Solve It?"

<img src="images/package-front.jpeg" alt="Instant Insanity package" style="width: 25%;">

### Voiceover: lasted updated 2025-04-04
The 2024 Winning Solutions Instant Insanity package advertises 82,944 combinations.

## Robin Wilson, Gresham College: 82,944

See: [The Four Cubes Problem](https://youtu.be/sI7lX3M7OZo?si=ijtOwtu-WV1sY4lH)

### Scene

Play a clip from the YouTube video where Robin Wilson says there are 82,944 combinations.

### Voiceover: last updated 2025-04-11
The 2008 Gresham College lecture, The Four Cubes Problem
by Robin Wilson, Gresham Professor of Geometry, gives the number as 82,944.

## Tai-Danea Bradley, PBS Infinite Series: 82,944

See: [Instant Insanity Puzzle](https://youtu.be/Lw1pF47N-0Q?si=WX3NlcEhPNLdY6Na)

### Scene

Play a clip from the YouTube video where Tai-Danea Bradley says there are 82,944 combinations.

### Voiceover: last updated 2025-04-11

The 2018 PBS Infinite Series video, Instant Insanity Puzzle  by Tai-Danea Bradley
also gives the number as 82,944.

## F. de Carteblanche, Eureka: 41,472

See: [The Coloured Cubes Problem
](https://onedrive.live.com/?authkey=%21AG4HZAmpHD0fSWs&cid=BF134020839BC69B&id=BF134020839BC69B%21210&parId=BF134020839BC69B%21147&o=OneUp)

### Scene

Show the statement from the 1947 paper.

### Voiceover: last updated 2025-04-04

The 1947 Eureka paper, The Coloured Cubes Problem
by F. de Carteblanche which gives the elegant graph theory solution to the problem states that if you pick a random arrangement
of the cubes your chance of it being a solution is 1 in 41,472.
At that time, the puzzle was called the Tantalizer.
F. de Carteblanche was the pseudonym of a group of four Cambridge mathematics students which included the future eminent graph theorist
Bill Tutte.

## Which Number is Correct?

### Scene

Show the puzzle box (82,944) and the Carteblanche paper (41,472) side-by-side, highlighting the two numbers.

### Voiceover: last updated 2025-04-11

Neither the original Cambridge paper nor the YouTube videos actually show us how they arrived at their numbers.
Let's apply some combinatorics and compute the answer ourselves.

## Combination and Solutions

### Scene

Show a vertical stack solution and the sentence from the 1947 Carteblache paper that says:

> The problem is to stack these cubes in a vertical pile (thus forming a square prism) in such a way that each of the four vertical faces of this pile contains all four colours.

Show a horizontal row solution and the sentence from the back of the puzzle box that says:

> Can you arrange the cubes in a row or stack so that all four colors appear on each side?

Maybe split the screen into left and right frames and show a vertical stack on the left and a horizontal row on the right, with the textual quotes below them.

### Voiceover: last updated 2025-04-14

First we need to precisely define what we mean by a combination of the cubes and a solution of the puzzle.
The 1947 Carteblanche paper describes the goal of the puzzle as follows:

The problem is to stack these cubes in a vertical pile (thus forming a square prism) in such a way that each of the four vertical faces of this pile contains all four colours.

The puzzle box asks:

Can you arrange the cubes in a row or stack so that all four colors appear on each side?

We'll call any vertical or horizontal arrangement of the four cubes a combination and any combination that solves the puzzle a solution.

Clearly, we'll get the same number of combinations whether we arrange the cubes vertically or horizontally since we can easily convert one type of arrangement into the other.
In this video, we'll usually arrange the cubes in a horizonal row.

## The Probability of a Solution

### Scene

Show the following mathematical formula. I'll generate a 720x1440 pixel PNG image.

\begin{align*}
N &: \mathrm{the\ number\ of\ combinations} \\
S &: \mathrm{the\ number\ of\ solutions} \\
P &: \mathrm{the\ probability\ of\ a\ solution} \\
\\
P &= \frac{S}{N} = \frac{1}{41,472}
\end{align*}

### Voiceover: last updated 2025-04-14

Carteblanche makes the following claim.

The chance of obtaining the solution by a random arrangement of the cubes is only 1 in 41,472.

Let N be the total number of possible cube combinations and let S be the total number of solutions.
Let P be the probability that a randomly chosen combination is a solution.
The probability P is therefore the number of solutions S divided by the number of combinations N.

Let's compute N and S and see what we get for P.

## Generating all Possible Combinations

### Scene

Animate a random cube combination generator.
This could be like one of those lottery machines that scrambles numbered balls and pumps them out one at a time to produce the winning number.
Alternatively, it could be a physics simulation of dropping the four cubes and then lining them up in a horizontal row.

### Voiceover: last updated 2025-04-14

To generate all possible combinations, consider the following random procedure.
Put the four cubes in a bag, shake it, draw the cubes out one at a time, and, without looking at them, arrange them in a row.
If we repeat this procedure often enough it will eventually produce all possible combinations.

How many combinations can we produce?

Observe that we can fully describe any combination by first giving the order of the cubes,
and then giving the orientation of each cube.

## Naming the Cubes and Their Faces

### Scene

The purpose of this scene is to assign names to the cubes and their faces so we can talk about them more easily.
We'll use the naming scheme on p9 of the Carteblanche article.

Start by showing a horizontal row of the four cubes.
The perspective should be chosen so that the front, right, and top faces are visible.

Fade in the numbers from 1 to 4 to name the cubes.

Fade in the names x, y, and z on the front, right, and top faces of each cube.

Animate unfolding the cubes and add the names x', y', and z' to the back, left, and bottom faces.

Animate rearranging the faces of each flattened cube into a vertical column of six faces in the order x, x', y, y', z, z'.

Animate moving the faces into an array with six rows (one per face) and four columns (one per cube).

Fade in the table row and column names to produce a table like that on page 9 of Carteblache but using the actual layout of the cubes in the current Winning Moves puzzle.
Here is the table for the puzzle.

| Face | 1 | 2 | 3 | 4 |
|------|---|---|---|---|
| x (front)   | G | R | W | B |
| x' (back)   | W | G | R | R |
| y (right)   | B | B | W | G |
| y' (left)   | R | B | B | W |
| z (top)     | R | W | G | B |
| z' (bottom) | R | G | R | W |

### Voiceover: last updated 2025-04-14

The task of describing and counting the number of combinations will be easier if we assign short symbolic names to the cubes and their faces.

We'll use the naming scheme defined on p9 of the Carteblanche article.

In this scheme, we pick some arbitrary order for the cubes and name them 1, 2, 3, and 4.

We then pick some arbitrary orientation for each cube and name the front, right, and top faces x, y, and z.
We name their opposite faces x', y', and z'.

For example, face 3y' is the name of the left face of cube 3 which is blue.

## Counting the Number of Cube Orderings

### Scene

The purpose of this scene is to count the number of cube orderings.

A nice way to visualize this is to create a horizontally oriented tree.

Put the root node on the left. It represents and empty sequence of cubes.

We need to pick the first cube in the order.
There are four choices so draw 4 arcs from the root to each of the choices.
We could draw the cube with its name or just the name.

Next we choose the second cube.
There are now only three possible choices so draw three arcs from the first cube to each of the three possible second cubes.

Proceed until all four cubes have been chosen.
The final tree has 24 leaf nodes.

Animate counting the leaf nodes by writing the numbers 1 to 24 to the right of each root node.

### Voiceover: last updated 2025-04-14

We're now ready to count the total number of possible cube orderings.
We'll build up each ordering from left to right and keep track of our progress in a decision tree.
Each cube ordering requires us to make four choices.
We have four choices for the first cube.
After we have made the first choice there are three cubes left so for each choice of the first cube we have three possibilities left for the second cube.
Similarly, for each ordering of the first two cubes we have two possibilities left for the third cube.
Finally, for each ordering of the first three cubes we only have one possibility left for the four cube.

Count all the cube orderings. There are 24 of them.

## Factorials

### Scene

The purpose of this scene is to make contact with the concept of permuations and the factorial function.

Show an image of the math formula for $n!$.

$$n! = n \times (n-1) \times \dots \times 2 \times 1$$ 

Then fade that into the case $n = 4$ giving the answer $24$.

$$4! = 4 \times 3 \times 2 \times 1 = 24$$ 

### Voiceover: last updated 2025-04-14

Combinatorics is the branch of mathematics that tells us how to count things.

One of the most important cases in combinatorics is how to count the number of ways that you can arrange a set of distinct objects in some order.
An arrangement of a set of distinct objects in some order is called a permuation of that set.

Combinatorics tells us that there are n factorial permuations of a set of n distinct objects.
To see this, observe that there are n ways to pick the first object, then n minus 1 ways to pick the second, and so on.

This is exactly how we counted the number of cube orders.
There are four cubes and four factorial equals twenty-four.

## The factorial Function in Python

This is an optional scene.

### Scene

The purpose of this scene is to make contact with combinatorics and Python.
Show the following Jupyter notebook cells.

### Voiceover: last updated 2025-04-14

Python has some built-in support for combinatorics.

Let's compute four factorial

In [1]:
4 * 3 * 2 * 1


24

### Voiceover: last updated 2025-04-14

Python has a built-in factorial function.
Import it from the math package.

In [2]:
from math import factorial

number_of_cube_orderings = factorial(4)

print(number_of_cube_orderings)

24


## Counting the Number of Cube Orientations

### Scene

The purpose of this scene is to count the number of cube orientations.
We'll build up each orientation by first choosing the top face and then choosing the front face.

Start with an animation showing a cube. 
Rotate the cube slowly about each axis so that we can see the face names x, y, z, x', y', and z'.

Next we'll build up a decision tree from left to right.
The leftmost node is the tree root which represents no choices yet.

We'll choose the top face first.
Draw six lines from the root to nodes that represent the six possible top faces x, y, z, x', y', and z'.

Then draw four lines from each of the six possible top face nodes to four more nodes that represent the possible front faces.

The following table summarizes the possible choices for the front face given a choice for the top face.
| Top | Front |
|-----|-------|
| x | y, z, y', z' |
| x' | y, z, y', z' |
| y | x, z, x', z' |
| y' | x, z, x', z' |
| z | x, y, x', y' |
| z | x, y, x', y' |

Finally, count the number of the leaf nodes.
There are 24. Draw each cube to the right of the corresponding leaf node.

Here is a table of the orientations.

| Number | Top | Front |
|--------|-----|-------|
| 1 | x | y |
| 2 | x | z |
| 3 | x | y' |
| 4 | x | z' |
| 5 | x' | y |
| 6 | x' | z |
| 7 | x' | y' |
| 8 | x' | z' |
| 9 | y | x |
| 10 | y | z |
| 11 | y | x' |
| 12 | y | z' |
| 13 | y' | x |
| 14 | y' | z |
| 15 | y' | x' |
| 16 | y' | z |
| 17 | z | x |
| 18 | z | y |
| 19 | z | x' |
| 20 | z | y' |
| 21 | z' | x |
| 22 | z' | y |
| 23 | z' | x' |
| 24 | z' | y' |

### Voiceover: last updated 2025-04-14

How many ways can a single cube be oriented?
We can specify an orientation by saying which face is on top and which face is in front.
Let's count the number of orientations.

We'll use another basic theorem of combinatorics. 
If we have two sets of things, say A containing n things and B containing m things
then the number of ways we can pair things from A with things from B is n times m.

A cube has 6 faces. Therefore, there are 6 ways to pick the top face.

Having picked the top face, we can pick any of the four adjacent faces as the front face by spinning the cube about its vertical axis.
We can do this in four ways.

Therefore, the total number of cube orientations is six times four which equals twenty-four.

## Optional Python Code

In [3]:
number_of_top_face_choices = 6

number_of_front_face_choices = 4

number_of_cube_orientations = number_of_top_face_choices * number_of_front_face_choices

print(number_of_cube_orientations)

24


## Counting the Number of Combinations

### Scene

The purpose of this scene is to compute the total number of combinations $N$.
We'll use the previously computed values for the number of cube orderings and orientations.

Show an image of the following math formula:

\begin{align*}
N 
    &= \mathrm{number\ of\ cube\ orderings} \times \mathrm{number\ of\ cube\ orientations}^4 \\
    &= 24 \times 24^4 \\
    &= 24^5 \\
    &= 7,962,624
\end{align*}

### Voiceover: last updated 2025-04-14

A combination is fully specified by giving the ordering of the four cubes and the orientation of each of cube.
There are 24 orderings of the cubes.
Each cube has 24 orientations.
Apply the basic combination rule again to get the total number of combinations.
Multiply the number of cube orderings and the number of cube orientations multiplied by itself four times
which is the same as raising the number of cube orientations to the power 4.

We get a whopping total of 7,962,624 combinations which is much bigger than the number of combinations, namely 82,944, claimed on the box and in the YouTube videos.
Why the difference?

The answer is that although there are in fact 7,962,624 distinct cube combinations when we randomly draw them out of a bag, we can be more intelligent in our search for a solution
by skipping combinations that are essentially equivalent to combinations we have already considered.

## Optional Python Code

In [4]:
N = number_of_combinations = number_of_cube_orderings * number_of_cube_orientations ** 4

print(N)

7962624


## When Are Two Combinations Essentially Equivalent?

### Scene

The purpose of this scene is to introduce the notion of essentially equivalent combinations by first considering the cube ordering.

Line up the four cubes in a solution.
The following combination is a solution:

$$
(1zx', 2zx', 3zy', 4zx')
$$

Animate the action of swapping the positions of any two cubes and noting that we still have a solution.
Do a few swaps, e.g. swap cubes 1 and 2, 2 and 4, etc.
If we started with a solution then we still have a solution after the swaps.

### Voiceover: last updated 2025-04-14

We can specify a combination by giving a cube ordering and an orientation for each cube.
An orientation is given by specifying the top and front faces.
Consider the combination given by the four-tuple (one z x', two z x', three z y', four z x').
This combination is a solution.

Now create a new combination by swapping the positions of any two cubes, say one and two, without changing their orientations.
The new combination is given by the four-tuple (two z x', one z x', three z y', four z x').
Since each side of the row contained four different colours before the swap, it will still contain four different colours after the swap.
Therefore the new combination is also a solution.
We can achieve any cube ordering by swapping pairs of cubes.
In fact, given any solution, it belongs to a family of twenty-four solutions that are related to eachother by swaps.

Swapping the positions of the cubes without changing their orientations is an example of what is called a transformation of the set
of all combinations since it sends combinations to combinations.
A set of related transformations is said to form a group.
Furthermore, swaps are said to be symmetries of the set of all solutions since they send solutions to solutions.
Group theory is the branch of mathematics that tells us about transformations and symmetries.

It is natural to regard solutions that differ only in the cube ordering as being essentially equivalent since we can transform one into the other by a sequence of swaps.
We can therefore use this notion of equivalence to cut down the size of our search space by a factor of twenty-four.
Rather than examining all combinations for solutions, we can safely confine our attention to one fixed cube ordering, say the natural ordering one, two, three, four.

## The Number of Combinations Modulo Order

### Scene

The purpose of this scene is to compute the number of combinations that are essentially equivalent when we disregard cube orderings.

Display and image of the following mathematical formulas.

STOPPED HERE

### Voiceover

Let $O$ be the set of all order-changing stack symmetries, that is, operations that swap cube positions but not cube orientations.
As mentioned above, the number of stacks related to each other $O$ operations is 24.
We refer to this as the *size* of the set of related symmetry operations.
The set of order-changing symmetry operations evenly divides the set of all stacks into families of size 24 such that 
each stack within a family is related to every other stack in the family by some order-changing symmetry.
Compute the number of order-changing symmetry families.

In [5]:
size_O = 24
number_of_combinations_O = number_of_combinations // size_O

number_of_combinations_O

331776

We get 331,776 which is much bigger than 82,944.
Why the difference?
The reason is that we should regard more arrangements as being essentially the same.

Suppose we have a solution. Then we can easily get another solution simply by rotating the stack one quarter of a turn
about the long axis. Furthermore, we can do this four times before ending up with the arrangement we started from.
Therefore each solution belongs of a family of four essentially equivalent solutions.

**Animation #5 rotating the stack of cubes one-quarter turn about the long axis**

Let $Q$ be the set of these quarter-turn symmetry operations. Clearly, this operation does not change the order the cubes
and it sends one $O$ family to another. In fact, it groups the set the $O$ families into families of four related families.
We refer to these families of families as $OQ$ families.
Compute the number of $OQ$ families

In [6]:
size_Q = 4
number_of_combinations_OQ = number_of_combinations_O // size_Q

print(number_of_combinations_OQ)

82944


Progress! Now we understand where the number 82,944 comes from.
It is the number of stacks where we consider stacks related by operations in $O$ or $Q$ to be equivalent.

But what about 41,472?
The explanation is similar to the above.
We have yet another symmetry.

Suppose we have a solution. We can obtain another solution by rotating each cube by one half turn about its vertical axis.

**Animation #6: rotating each cube one-half turn perpendicular to the long axis**

Let $H$ be this half-turn operation.
We can apply the $H$ operation twice before returning to the starting arrangement.
Therefore, we need to further divide the number of arrangements by 2 to get the number of essentially distinct arrangements.

In [7]:
size_H = 2
number_of_combinations_OQH = number_of_combinations_OQ // size_H

print(number_of_combinations_OQH)

41472


Success! We now understand where the number 41,472 comes from. 
It is the number of essentially distinct stacks if we consider stacks related by $O$, $Q$, or $H$ operations to be equivalent.
Furthermore, as we will soon prove, there is exactly one $OQH$ family that solves the problem.
Carteblanche was right.

Relating this back to our variable $S$ we have

In [8]:
S = size_O * size_Q * size_H

print(N, S, N // S)

7962624 192 41472


There are around 8 million stacks but these are divided into families each consisting of 192 mutually equivalent stacks.
The total number of essentially distinct stacks is $41,472$.

## Brute-Force Search

We now know how many arrangements, namely 41,472, we'd have to check in order to find a solution.
Given enough time and patience, we could systematically generate every possible essentially distinct arrangement
and check if it was a solution. Mathematicians call this approach a *brute-force search*.

How long would a brute-force search take?
Suppose a human was doing the search. As a rough estimate, suppose it takes 1 second to rotate a cube and 1 second to check
if the arrangement is a solution.
Let's compute the total time.

In [9]:
seconds_per_combination = 5
total_seconds = number_of_combinations_OQH * seconds_per_combination

print(total_seconds)

207360


In [10]:
seconds_per_minute = 60
total_minutes = total_seconds / seconds_per_minute

print(total_minutes)

3456.0


In [11]:
minutes_per_hour = 60
total_hours = total_minutes / minutes_per_hour

print(total_hours)

57.6


In [12]:
hours_per_day = 24
total_days = total_hours / hours_per_day

print(total_days)

2.4


Therefore, a human would find the solution if they worked nonstop for 2.4 days and made no errors.
This is called the *worst case* time since it assumes that you are unlucky and have to generate all the combinations
before you find the solution.

Clearly, this puzzle is challenging which accounts for its popularity.
The package states that over 20 million copies have been sold!

## Next

How long would a computer take to find the solution? 
One of us actually tried this in 1967 using a Fortran 4 program running on their high school's IBM 1130 computer.
Next, we'll write some Python code to perform the brute-force search and then optimize it for speed.